In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
dataset=pd.read_csv("iris dataset.csv")
dataset.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [3]:
x=dataset.iloc[0:, 0:4]
y=dataset.iloc[0:, 4:]
x=x.values
y=y.values


In [4]:
from sklearn.preprocessing import LabelEncoder

label=LabelEncoder()
y=label.fit_transform(y)

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/preprocessing/_label.py:251: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [5]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [6]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y, train_size=0.8)

In [7]:
from sklearn.linear_model import LogisticRegression
log=LogisticRegression()
modellog=log.fit(x_train, y_train)
rlog=modellog.score(x_train,y_train)

/srv/conda/envs/notebook/lib/python3.6/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [8]:
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier(criterion='entropy', random_state=0)
modeldt=dt.fit(x_train,y_train)
rdt=modeldt.score(x_train,y_train)

In [9]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators=10, criterion='entropy', random_state=0)
modelrf=rf.fit(x_train, y_train)
rrf=modelrf.score(x_train, y_train)

In [10]:
from sklearn.svm import SVC
svc=SVC()
modelsvc=svc.fit(x_train, y_train)
rsvc=modelsvc.score(x_train, y_train)

In [11]:
from sklearn.naive_bayes import GaussianNB
nb=GaussianNB()
modelnb=nb.fit(x_train, y_train)
rnb=modelnb.score(x_train, y_train)

In [12]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=3, metric='euclidean')
modelknn=knn.fit(x_train, y_train)
rknn=modelknn.score(x_train, y_train)

In [13]:
print('rscore of Logistic_regression is = ', rlog)
print()
print('rscore of decision_tree is = ', rdt)
print()
print('rscore of Random_Forest is = ', rrf)
print()
print('rscore of SVC is = ', rsvc)
print()
print('rscore of Naive_Bayes is = ', rnb)
print()
print('rscore of KNN is = ', rknn)
print()

rscore of Logistic_regression is =  0.9916666666666667

rscore of decision_tree is =  1.0

rscore of Random_Forest is =  1.0

rscore of SVC is =  0.975

rscore of Naive_Bayes is =  0.95

rscore of KNN is =  0.9583333333333334



# finding Cross_Val_Score for all models to check which model works well for this particular dataset

In [14]:
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import cross_val_score

print("cross_val_score of Logistic_regression is = ", '\n', cross_val_score(modellog, x,y, cv=10, scoring='accuracy'))
print()
print("cross_val_score of Decision_Tree is = ", '\n', cross_val_score(modeldt, x,y, cv=10, scoring='accuracy'))
print()
print("cross_val_score of Random_forest is = ", '\n', cross_val_score(modelrf, x,y, cv=10, scoring='accuracy'))
print()
print("cross_val_score of Support_Vector is = ", '\n', cross_val_score(modelsvc, x,y, cv=10, scoring='accuracy'))
print()
print("cross_val_score of Naive_Bayes is = ", '\n', cross_val_score(modelnb, x,y, cv=10, scoring='accuracy'))
print()
print("cross_val_score of KNN is = ", '\n', cross_val_score(modelknn, x,y, cv=10, scoring='accuracy'))
print()


cross_val_score of Logistic_regression is =  
 [1.         0.93333333 1.         1.         0.93333333 0.93333333
 0.93333333 1.         1.         1.        ]

cross_val_score of Decision_Tree is =  
 [1.         0.93333333 1.         0.93333333 0.93333333 0.86666667
 0.93333333 1.         1.         1.        ]

cross_val_score of Random_forest is =  
 [1.         0.93333333 1.         0.93333333 0.93333333 0.93333333
 0.93333333 0.93333333 1.         1.        ]

cross_val_score of Support_Vector is =  
 [1.         0.93333333 1.         1.         1.         0.93333333
 0.93333333 0.93333333 1.         1.        ]

cross_val_score of Naive_Bayes is =  
 [0.93333333 0.93333333 1.         0.93333333 0.93333333 0.93333333
 0.86666667 1.         1.         1.        ]

cross_val_score of KNN is =  
 [1.         0.93333333 1.         0.93333333 0.86666667 1.
 0.93333333 1.         1.         1.        ]



# Finding mean value for the above cross_val_score to check which model works fine for this particular dataset (IRIS Dataset)

In [15]:

print("mean value of cross_val_score of Logistic_regression is = ", cross_val_score(modellog, x,y, cv=10, scoring='accuracy').mean())
print()
print("mena value of cross_val_score of Decision_Tree is = ", cross_val_score(modeldt, x,y, cv=10, scoring='accuracy').mean())
print()
print("mean value of cross_val_score of Random_forest is = ", cross_val_score(modelrf, x,y, cv=10, scoring='accuracy').mean())
print()
print("mean value of cross_val_score of Support_Vector is = ", cross_val_score(modelsvc, x,y, cv=10, scoring='accuracy').mean())
print()
print("mean value of cross_val_score of Naive_Bayes is = ", cross_val_score(modelnb, x,y, cv=10, scoring='accuracy').mean())
print()
print("mean value of cross_val_score of KNN is = ", cross_val_score(modelknn, x,y, cv=10, scoring='accuracy').mean())
print()

mean value of cross_val_score of Logistic_regression is =  0.9733333333333334

mena value of cross_val_score of Decision_Tree is =  0.96

mean value of cross_val_score of Random_forest is =  0.96

mean value of cross_val_score of Support_Vector is =  0.9733333333333334

mean value of cross_val_score of Naive_Bayes is =  0.9533333333333334

mean value of cross_val_score of KNN is =  0.9666666666666666



## from the above it is evident that "Logistic Regression" model works well for this dataset, as the "mean cross_val_score"  of "logistic regression" is high (0.9733) after splitting the dataset in to 10 partions for training and testing randomly.